crates/oxc_linter/src/rules/eslint/dot_notation.rs

```diff
@@ -0,0 +1,343 @@
+    "new",
+    "null",
+    "package",
+    "private",
+    "protected",
+    "public",
+    "return",
+    "short",
+    "static",
+    "super",
+    "switch",
+    "synchronized",
+    "this",
+    "throw",
+    "throws",
+    "transient",
+    "true",
+    "try",
+    "typeof",
+    "var",
+    "void",
+    "volatile",
+    "while",
+    "with",
+];
+lazy_static! {
+    static ref VALID_IDENTIFER: Regex = Regex::new(r"^[a-zA-Z_$][a-zA-Z0-9_$]*$").unwrap();
+}
+
+impl Rule for DotNotation {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let mut cfg = DotNotationConfig::default();
+
+        if let Some(config) = value.get(0) {
+            if let Some(val) = config.get("allowPattern").and_then(serde_json::Value::as_str) {
+                cfg.allow_pattern = RegexBuilder::new(val).build().ok();
+            }
+            if let Some(val) = config.get("allowKeywords").and_then(serde_json::Value::as_bool) {
+                cfg.allow_keywords = val;
+            }
+        }
+
+        Self(Box::new(cfg))
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::MemberExpression(node) = &node.kind() {
+            match &node {
+                MemberExpression::ComputedMemberExpression(expr) => {
+                    let inner_expr =
+                        if let Expression::ParenthesizedExpression(inner_expr) = &expr.expression {
+                            &inner_expr.expression
+                        } else {
+                            &expr.expression
+                        };
+
+                    if let Some((before, after)) = match &inner_expr {
+                        Expression::NullLiteral(x) => Some((x.to_string(), x.to_string())),
+                        Expression::BooleanLiteral(x) => Some((x.to_string(), x.to_string())),
+                        Expression::StringLiteral(x) => Some((x.to_string(), x.to_string())),
+                        Expression::TemplateLiteral(x) => {
+                            if x.expressions.len() == 0 && x.quasis.len() == 1 {
+                                let s = x.quasis[0].value.cooked.unwrap().into_string();
+                                Some((s.clone(), s))
+                            } else {
+                                None
+                            }
+                        }
+                        _ => None,
+                    } {
+                        if VALID_IDENTIFER.is_match(&after)
+                            && (self.0.allow_keywords || !KEYWORDS.contains(&after.as_str()))
+                            && !(self.0.allow_pattern.is_some()
+                                && self.0.allow_pattern.as_ref().unwrap().is_match(&after))
+                        {
+                            ctx.diagnostic(dot_notation_use_dot_diagnostic(
+                                expr.span, &before, &after,
+                            ))
+                        }
+                    }
```

Comment:
no need to unwrap the regex pattern, you can use `if-let` blocks. I would also recommend flattening this out a bit with early returns to make it a little easier to follow. taking other suggestions into account:

```suggestion
                if !is_valid_identifier(&value) {
                    return;
                }
                if !self.0.allow_keywords && KEYWORDS.contains(&value.as_str()) {
                    return;
                }

                if let Some(pattern) = &self.0.allow_pattern {
                    if pattern.is_match(&value) {
                        return;
                    }
                }

                ctx.diagnostic(dot_notation_use_dot_diagnostic(expr.span, &value, &value))
```